src/darwin/Framework/CHIP/MTRBaseDevice.mm

```diff
@@ -908,8 +936,162 @@ - (void)readAttributesWithEndpointID:(NSNumber * _Nullable)endpointID
+                    MTRAttributePathKey : [[MTRAttributePath alloc] initWithPath:attribPath],
+                    MTRDataKey : aData.GetDecodedObject()
+                }];
+            };
+
+            auto onEventSuccessCb = [resultArray](const app::ConcreteClusterPath & clusterPath, const EventId aEventId,
+                                        const MTRDataValueDictionaryDecodableType & aData) {
+                app::ConcreteEventPath eventPath(clusterPath.mEndpointId, clusterPath.mClusterId, aEventId);
+                [resultArray addObject:@ {
+                    MTREventPathKey : [[MTREventPath alloc] initWithPath:eventPath],
+                    MTRDataKey : aData.GetDecodedObject()
+                }];
+            };
+
+            auto onFailureCb = [resultArray, interactionStatus](
+                                   const app::ConcreteClusterPath * clusterPath, const uint32_t aValueId, CHIP_ERROR aError) {
+                if (clusterPath) {
+                    app::ConcreteAttributePath attribPath(clusterPath->mEndpointId, clusterPath->mClusterId, aValueId);
+                    [resultArray addObject:@ {
+                        MTRAttributePathKey : [[MTRAttributePath alloc] initWithPath:attribPath],
+                        MTRErrorKey : [MTRError errorForCHIPErrorCode:aError]
+                    }];
+                } else {
+                    // This will only happen once per read interaction, and
+                    // after that there will be no more calls to onFailureCb or
+                    // onSuccessCb.
+                    *interactionStatus = aError;
+                }
+            };
+
+            std::vector<app::AttributePathParams> attributePathParamsList;
+            std::vector<app::EventPathParams> eventPathParamsList;
+
+            if (attributePaths != nil) {
+                for (uint8_t i = 0; i < [attributePaths count]; i++) {
+                    chip::EndpointId endpointId = kInvalidEndpointId;
+                    chip::ClusterId clusterId = kInvalidClusterId;
+                    chip::AttributeId attributeId = kInvalidAttributeId;
+
+                    if ([attributePaths[i] endpoint]) {
+                        endpointId = static_cast<chip::EndpointId>([[attributePaths[i] endpoint] unsignedShortValue]);
+                    }
+
+                    if ([attributePaths[i] cluster]) {
+                        clusterId = static_cast<chip::ClusterId>([[attributePaths[i] cluster] unsignedLongValue]);
+                    }
+
+                    if ([attributePaths[i] attribute]) {
+                        attributeId = static_cast<chip::AttributeId>([[attributePaths[i] attribute] unsignedLongValue]);
+                    }
+                    attributePathParamsList.push_back(app::AttributePathParams(endpointId, clusterId, attributeId));
+                }
+            }
+
+            if (eventPaths != nil) {
+                for (uint8_t i = 0; i < [eventPaths count]; i++) {
```

Comment:
Similar comments here as for attributes.

Reply:
I modified